Dynamic Module Import

suwc opened the pull request:
This PR adds support for `import()` dynamic module import sematic.

Per https://github.com/tc39/proposal-dynamic-import:

"A call to `import(specifier)` returns a promise for the module namespace object of the requested module, which is created after fetching, instantiating, and evaluating all of the module's dependencies, as well as the module itself.

"Here specifier will be interpreted the same way as in an import declaration (i.e., the same strings will work in both places). However, while specifier is a string it is not necessarily a string literal; thus code like import(`./language-packs/${navigator.language}.js`) will work—something impossible to accomplish with the usual import declarations.

"`import()` is proposed to work in both scripts and modules. This gives script code an easy asynchronous entry point into the module world, allowing it to start running module code."

This PR includes following changes:
	- Update parser and bytecode generator to support `import()` sematic in module and in script
	- Add new bytecode `ImportCall`
	- Add runtime function for `import()` that:
		○ Uses caller from stack to look up module record or source context that are associated with the module or script from which `import()` is called
		○ Requests host to load target module source file (gets module record in return)
		○ Creates promise unless the module record has one
		○ Resolves/rejects promise if appropriates
		○ Returns promise
	- Add new host callback (`FetchImportedModuleFromScript`) for fetching imported module from script (accepts source context)
	- Add `promiseCapability` field to module record class
	- Update `SourceTextModuleRecord`'s methods to accept callback from host and to handle dynamically imported module and its promise capability
	- Update exception checks and assertions to cover new usage scenario of importing and evaluating module code with active script
Add unit tests for dynamic import functionality

suwc commented:
@Microsoft/chakra-es @Microsoft/chakra-developers, please help code review. Thanks!

agarwal-sandeep commented on lib/Jsrt/ChakraCore.h:
```diff
@@ -70,6 +71,21 @@ typedef JsErrorCode(CHAKRA_CALLBACK * FetchImportedModuleCallBack)(_In_ JsModule
 /// <returns>
 ///     true if the operation succeeded, false otherwise.
 /// </returns>
+typedef JsErrorCode(CHAKRA_CALLBACK * FetchImportedModuleFromScriptCallBack)(_In_ DWORD_PTR dwReferencingSourceContext, _In_ JsValueRef specifier, _Outptr_result_maybenull_ JsModuleRecord* dependentModuleRecord);
```
>DWORD_PTR [](start = 82, length = 9)

Use JsSourceContext instead of DWORD_PTR

suwc replied:
Updated in 2nd iteration.

---
In reply to: [114836637](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114836637) [](ancestors = 114836637)

agarwal-sandeep commented on test/es6/dynamic-module-functionality.js:
```diff
@@ -0,0 +1,359 @@
+                `testDynamicImport(
+                    ()=>{
+                        var getName = ()=>{ return 'ModuleSimpleExport'; };
+                        return import( getName() + '.js');
```
>import [](start = 31, length = 6)

What if a non-string is passed to import? like import() or import({}) etc. Can we add test cases covering error scenarios?

suwc replied:
Added new test dynamic-module-import-specifier.js.
More tests to come.

---
In reply to: [114840008](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114840008) [](ancestors = 114840008)

agarwal-sandeep commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,82 @@ namespace Js
+            {
+                Js::JavascriptError * error = scriptContext->GetLibrary()->CreateURIError();
+                Var value = JavascriptString::NewCopySz(moduleName, scriptContext);
+                JavascriptOperators::OP_SetProperty(error, PropertyIds::message, value, scriptContext);
```
>OP_SetProperty [](start = 37, length = 14)

Should you use JavascriptError::SetErrorMessageProperties instead?

suwc replied:
Updated in 2nd iteration.

---
In reply to: [114844985](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114844985) [](ancestors = 114844985)

agarwal-sandeep commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,82 @@ namespace Js
+        }
+        else
+        {
+            Assert(false);
```
>Assert [](start = 12, length = 6)

nit: Using AssertMsg with a message would be more useful

suwc replied:
Good point. This line is removed along with stack walker call though.

---
In reply to: [114845379](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114845379) [](ancestors = 114845379)

boingoing commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -983,4 +1035,41 @@ namespace Js
+            promiseCapability = JavascriptPromise::NewPromiseCapability(scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
+        }
+
+        Var resolveOrRejectVar = toResolve ? promiseCapability->GetResolve() : promiseCapability->GetReject();
```
So `toResolve` indicates the promise should be resolved instead of rejected? Nitpicky, but I think I would prefer this to be named `isResolve`, `isReject`, `success`, or something along those lines.

tcare replied:
+1, I also found this hard to read.

---
In reply to: [114906734](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114906734) [](ancestors = 114906734)

suwc replied:
Changed to 'isResolve'

---
In reply to: [114906734](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114906734) [](ancestors = 114906734)

boingoing commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,86 @@ namespace Js
+                if (promiseCapability == nullptr)
+                {
+                    BEGIN_TRANSLATE_TO_HRESULT(static_cast<ExceptionType>(ExceptionType_OutOfMemory | ExceptionType_StackOverflow));
+                    promiseCapability = JavascriptPromise::NewPromiseCapability(scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
```
Does the spec call-out that we should be storing the `PromiseCapabilityRecord` here or is this more of an implementation detail? I would suspect it should be simpler to store a `JavascriptPromise*` here and call `JavascriptPromise::Resolve()` or `JavascriptPromise::Reject()` on it when it's time to resolve/reject it.

Looks like we're storing `promiseCapability` here to avoid casting the promise into a `JavascriptPromise*` because `JavascriptPromiseCapability::promise` is a `Var`? I think that cast should be safe to do since we pass `%Promise%` to `JavascriptPromise::NewPromiseCapability` which is guaranteed to return something that's a `JavascriptPromise*` because the return object is constructed via `JavascriptLibrary::CreatePromise` and we aren't doing a super call. Actually might be even simpler than that since we don't even need to call an executor function for this promise, right? So we could just call `library->CreatePromise()` and pass the result to `JavascriptPromise::InitializePromise`.

Might make sense to add a helper to `JavascriptPromise` itself

```c++
    JavascriptPromise* JavascriptPromise::CreateEnginePromise(ScriptContext* scriptContext)
    {
        // Unused
        JavascriptPromiseResolveOrRejectFunction* resolve;
        JavascriptPromiseResolveOrRejectFunction* reject;

        JavascriptPromise* promise = library->CreatePromise();
        InitializePromise(promise, &resolve, &reject, scriptContext);

        return promise;
    }
```

I think it would be simpler to resolve the promise at least if we stored the promise itself instead of the capability record. We would also avoid constructing a bunch of objects and making a bunch of calls by constructing the promise as the above helper.

suwc replied:
Good idea. Updated in 2nd iteration.

---
In reply to: [114912538](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114912538) [](ancestors = 114912538)

tcare commented on lib/Runtime/Language/i386/StackFrame.cpp:
```diff
@@ -51,9 +51,13 @@ X86StackFrame::InitializeByReturnAddress(void * returnAddress, ScriptContext* sc
-    this->codeAddr = this->GetReturnAddress();
-    this->frame = (void **)this->frame[0];
+    if (frame != nullptr)
+    {
```
Which situation needs this and why only x86?

suwc replied:
Only with nested enter_script, which I have removed.

---
In reply to: [114913205](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114913205) [](ancestors = 114913205)

tcare commented on lib/Parser/Parse.cpp:
```diff
@@ -2436,19 +2436,26 @@ bool Parser::IsImportOrExportStatementValidHere()
+    if (m_token.tk == tkLParen)
+    {
+        ParseNodePtr specifier = ParseTerm<buildAST>();
+        return CreateCallNode(knopCall, CreateNodeWithScanner<knopImport>(), specifier);
```
>CreateCallNode [](start = 15, length = 14)

Since we parse this like a call, I would like to make sure we have sufficient coverage for parsing cases where import() could be mistaken for a function call, e.g. with multiple parameters, spread, etc.

boingoing replied:
This is a good point

suwc replied:
Added dynamic-module-import-specifier.js
More to come.

---
In reply to: [114915006](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114915006) [](ancestors = 114915006)

tcare commented on lib/Runtime/ByteCode/ByteCodeEmitter.cpp:
```diff
@@ -10803,6 +10803,14 @@ void Emit(ParseNode *pnode, ByteCodeGenerator *byteCodeGenerator, FuncInfo *func
+        else if (pnode->sxCall.pnodeTarget->nop == knopImport)
+        {
+            ParseNodePtr args = pnode->sxCall.pnodeArgs;
+            Emit(args, byteCodeGenerator, funcInfo, false);
```
>args [](start = 17, length = 4)

Since we're piggybacking off a call node, perhaps we want to assert some restrictions e.g. single parameter

suwc replied:
Added assertion in 2nd iteration.

---
In reply to: [114915361](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114915361) [](ancestors = 114915361)

curtisman commented on lib/Common/Exceptions/ExceptionCheck.cpp:
```diff
@@ -58,7 +58,8 @@ void ExceptionCheck::SetHandledExceptionType(ExceptionType e)
     if(!(e == ExceptionType_None ||
          e == ExceptionType_DisableCheck ||
-         !JsUtil::ExternalApi::IsScriptActiveOnCurrentThreadContext() ||
+         (e & ExceptionType_OutOfMemory) == ExceptionType_OutOfMemory ||  // InitializeModuleRecord handles OOM during dynamic import
```
>InitializeModuleRecord [](start = 77, length = 22)

Why add this?  The point here is that if you are in script, you should be throwing the JS exception object instead of OutOfMemory.

suwc replied:
Removed along with nested EnterScript.

---
In reply to: [114929334](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114929334) [](ancestors = 114929334)

curtisman commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -7134,12 +7134,12 @@ namespace Js
             Js::Throw::FatalInternalError();
         }
-        return moduleRecordList->Item(moduleId);
+        return moduleRecordList->Item(moduleId - 1); // 0 == kmodGlobal
```
Why use kmodGlobal?  That is the value to distingish between the different global scopes (the real global scope of form global scope).  It has nothing to do with Javascript modules)
Please define a different constant to mean: this is not a module?  May be use the value -1 instead os you don't have to do this weird shifting?

curtisman commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -609,6 +645,15 @@ namespace Js
+        if (nullptr == this->childrenModuleSet)
+        {
+            ArenaAllocator* allocator = scriptContext->GeneralAllocator();
+            this->childrenModuleSet = (ChildModuleRecordSet*)AllocatorNew(ArenaAllocator, allocator, ChildModuleRecordSet, allocator);
```
>childrenModuleSet [](start = 18, length = 17)

Use Anew?

suwc replied:
Updated in 2nd iteration.

---
In reply to: [114930171](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114930171) [](ancestors = 114930171)

curtisman commented:
>         childrenModuleSet = nullptr;

childrenModuleSet is leaked?

---
Refers to: lib/Runtime/Language/SourceTextModuleRecord.cpp:77 in ea66233. [](commit_id = ea66233b2394329953901fc69aea3625b631d40a, deletion_comment = False)

curtisman commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,86 @@ namespace Js
+        SourceTextModuleRecord *moduleRecord = nullptr;
+
+        JavascriptFunction* pfuncCaller;
+        if (JavascriptStackWalker::GetCaller(&pfuncCaller, scriptContext) && pfuncCaller && pfuncCaller->IsScriptFunction())
```
Why do we need to use the stack walker here?    Didn't the interpreter just called this OP_ImportCall and know what the function is already?  Can't the interpreter pass it in?

suwc replied:
Updated interpreter and lowerer code to pass in function and removed stack walker call.

---
In reply to: [114930380](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114930380) [](ancestors = 114930380)

curtisman commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,86 @@ namespace Js
+        if (JavascriptStackWalker::GetCaller(&pfuncCaller, scriptContext) && pfuncCaller && pfuncCaller->IsScriptFunction())
+        {
+            FunctionBody* parentFuncBody = pfuncCaller->GetFunctionBody();
+            ModuleID id = parentFuncBody->GetModuleID();
```
The ModuleID here has no relationship with Javascript modules.
This is in reference to different global scope (either global, or in a  HTML form).
The type should rename ModuleRootID so it is not being confused.

curtisman commented:
>                 srcInfo->moduleID = moduleId;

moduleID in SRCINFO is unrelated to javascript modules.  It is probably wron't to store this here.

---
Refers to: lib/Runtime/Language/SourceTextModuleRecord.cpp:122 in ea66233. [](commit_id = ea66233b2394329953901fc69aea3625b631d40a, deletion_comment = False)

curtisman commented on lib/Runtime/Base/LeaveScriptObject.cpp:
```diff
@@ -63,7 +63,7 @@ namespace Js
 
             // Put any code that may raise an exception in OnScriptStart
-            scriptContext->GetThreadContext()->EnterScriptStart(entryExitRecord, doCleanup);
+            scriptContext->GetThreadContext()->EnterScriptStart(entryExitRecord, doCleanup, &this->isScriptActive);
```
are we trying to allow nested EnterScript without LeaveScript?  Why?

suwc replied:
This is to allow e.g. ImportCall() -> FetchImportedModule() -> InitializeModuleRecord().
When FetchImportedModule() from host calls into chakra, it goes through EnterScript-LeaveScript.
Non-nested EnterScript-LeaveScript resets threadContext->isScriptActive to false.


---
In reply to: [114931889](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114931889) [](ancestors = 114931889)

suwc commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -61,7 +61,7 @@ namespace Js
         // site. Host might hold a reference to the module as well after initializing the module.
         // In our implementation, we'll use the moduleId in bytecode to identify the module.
-        childModuleRecord->moduleId = scriptContext->GetLibrary()->EnsureModuleRecordList()->Add(childModuleRecord);
+        childModuleRecord->moduleId = scriptContext->GetLibrary()->EnsureModuleRecordList()->Add(childModuleRecord) + 1; // 0 == kmodGlobal
```
Found a better way to check module code so this shift is no longer needed.

---
In reply to: [114905856](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114905856) [](ancestors = 114905856)

suwc commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -230,15 +259,15 @@ namespace Js
-            if (!WasDeclarationInitialized() && isRootModule)
+            if (this->promiseCapability != nullptr || (!WasDeclarationInitialized() && isRootModule))
             {
                 // TODO: move this as a promise call? if parser is called from a different thread
```
This PR doesn't enable the case where parser is called from a different thread. Therefore I leave this TODO here for future enhancement.

---
In reply to: [114906209](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114906209) [](ancestors = 114906209)

suwc commented on lib/Common/Exceptions/ExceptionCheck.cpp:
```diff
@@ -58,7 +58,8 @@ void ExceptionCheck::SetHandledExceptionType(ExceptionType e)
          e == ExceptionType_DisableCheck ||
-         !JsUtil::ExternalApi::IsScriptActiveOnCurrentThreadContext() ||
+         (e & ExceptionType_OutOfMemory) == ExceptionType_OutOfMemory ||  // InitializeModuleRecord handles OOM during dynamic import
+        !JsUtil::ExternalApi::IsScriptActiveOnCurrentThreadContext() ||
```
Fixed in 2nd iteration

---
In reply to: [114900876](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114900876) [](ancestors = 114900876)

suwc commented:
>                 srcInfo->moduleID = moduleId;

Switched to using fscrIsModuleCode to detect ES6 module in ParseableFuncitonInfo::IsES6ModuleCode().
Removed shifting of moduleId by 1 and going back to starting from 0.
Will follow up SRCINFO->moduleID usage separately.

---
In reply to: [299373242](https://github.com/Microsoft/ChakraCore/pull/2913#issuecomment-299373242) [](ancestors = 299373242)

---
Refers to: lib/Runtime/Language/SourceTextModuleRecord.cpp:122 in ea66233. [](commit_id = ea66233b2394329953901fc69aea3625b631d40a, deletion_comment = False)

suwc commented:
>         childrenModuleSet = nullptr;

Added Adelete in 2nd iteration.

---
In reply to: [299368705](https://github.com/Microsoft/ChakraCore/pull/2913#issuecomment-299368705) [](ancestors = 299368705)

---
Refers to: lib/Runtime/Language/SourceTextModuleRecord.cpp:77 in ea66233. [](commit_id = ea66233b2394329953901fc69aea3625b631d40a, deletion_comment = False)

suwc commented on lib/Parser/Parse.cpp:
```diff
@@ -2436,19 +2436,26 @@ bool Parser::IsImportOrExportStatementValidHere()
     Assert(m_scriptContext->GetConfig()->IsES6ModuleEnabled());
     Assert(m_token.tk == tkIMPORT);
 
+    m_pscan->Scan();
```
Yes, updated in 2nd iteration.

---
In reply to: [114901862](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114901862) [](ancestors = 114901862)

suwc commented:
@dotnet-bot 
test Windows arm_release
test Windows x64_release
test Windows x86_release

akroshg commented on lib/Parser/Parse.cpp:
```diff
@@ -2435,12 +2435,37 @@ bool Parser::IsImportOrExportStatementValidHere()
+    ParseNodePtr specifier = ParseExpr<buildAST>(koplCma, nullptr, /* fAllowIn */FALSE, /* fAllowEllipsis */FALSE);
+    if (m_token.tk != tkRParen)
+    {
+        Error(ERRsyntax);
```
nit: proper error would be useful.

suwc replied:
Changing to 'ERRnoRparen'.

---
In reply to: [115324907](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115324907) [](ancestors = 115324907)

akroshg commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -1267,6 +1295,45 @@ JsErrorCode WScriptJsrt::FetchImportedModule(_In_ JsModuleRecord referencingModu
+// we are not doing any translation, just treat the specifier as fileName.
+// While this call will come back directly from runtime script or module code, the additional
+// task can be scheduled asynchronously that executed later.
+JsErrorCode WScriptJsrt::FetchImportedModuleFromScript(_In_ JsSourceContext dwReferencingSourceContext,
```
>FetchImportedModuleFromScript [](start = 25, length = 29)

wondering if this function can be refactored with above function. It seems most of the code is shared.

suwc replied:
updated in latest iteration

---
In reply to: [115327127](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115327127) [](ancestors = 115327127)

akroshg commented on lib/Jsrt/Core/JsrtContextCore.cpp:
```diff
@@ -120,6 +120,36 @@ HRESULT ChakraCoreHostScriptContext::FetchImportedModule(Js::ModuleRecordBase* r
+
+    HRESULT hr = NOERROR;
+    Js::JavascriptString* specifierVar = nullptr;
+    specifierVar = Js::JavascriptString::NewCopySz(specifier, GetScriptContext());
```
nit: combine these two lines?

suwc replied:
NA in latest iteration (file refactored)

---
In reply to: [115328093](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115328093) [](ancestors = 115328093)

akroshg commented on lib/Jsrt/Core/JsrtContextCore.cpp:
```diff
@@ -120,6 +120,36 @@ HRESULT ChakraCoreHostScriptContext::FetchImportedModule(Js::ModuleRecordBase* r
+    Js::JavascriptString* specifierVar = nullptr;
+    specifierVar = Js::JavascriptString::NewCopySz(specifier, GetScriptContext());
+    
+    if (FAILED(hr))
```
is this redundant check? where the hr is changed in above code?

suwc replied:
NA in latest iteration (function refactored).

---
In reply to: [115328224](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115328224) [](ancestors = 115328224)

akroshg commented on lib/Jsrt/Core/JsrtContextCore.cpp:
```diff
@@ -120,6 +120,36 @@ HRESULT ChakraCoreHostScriptContext::FetchImportedModule(Js::ModuleRecordBase* r
+        return hr;
+    }
+
+    JsModuleRecord dependentRecord = JS_INVALID_REFERENCE;
```
there is not much difference between this function and above function (FetchImportedModule). Did you consider creating a lambda and just stub relevant part in that?

suwc replied:
Refactored in latest iteration.

---
In reply to: [115328606](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115328606) [](ancestors = 115328606)

akroshg commented on lib/Jsrt/Core/JsrtContextCore.h:
```diff
@@ -177,6 +178,9 @@ class ChakraCoreHostScriptContext sealed : public HostScriptContext
     void SetFetchImportedModuleCallback(FetchImportedModuleCallBack fetchCallback) { this->fetchImportedModuleCallback = fetchCallback ; }
     FetchImportedModuleCallBack GetFetchImportedModuleCallback() const { return this->fetchImportedModuleCallback; }
 
+    void SetFetchImportedModuleFromScriptCallback(FetchImportedModuleFromScriptCallBack fetchCallback) { this->fetchImportedModuleFromScriptCallback = fetchCallback; }
```
Is this function allowed to be called multiple times? Is it good to assert that this->fetchImportedModuleFromScriptCallback is null when it is called?

suwc replied:
These functions are being called multiple times.
For example:

REQUIRE(JsSetModuleHostInfo(requestModule, JsModuleHostInfo_FetchImportedModuleCallback, Success_FIMC) == JsNoError);


---
In reply to: [115328835](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115328835) [](ancestors = 115328835)

akroshg commented:
>         case JsModuleHostInfo_NotifyModuleReadyCallback:

any update required in this function?

---
Refers to: lib/Jsrt/Core/JsrtCore.cpp:209 in db8d7fd. [](commit_id = db8d7fdc5b27af9f9814887a781d6e8c924e07b7, deletion_comment = False)

akroshg commented:
>         if ((this->GetGrfscr() & fscrIsModuleCode) == fscrIsModuleCode || this->GetModuleID() == kmodGlobal)

nit: can this be changed with IsES6ModuleCode()?

---
Refers to: lib/Runtime/Base/FunctionBody.cpp:3166 in db8d7fd. [](commit_id = db8d7fdc5b27af9f9814887a781d6e8c924e07b7, deletion_comment = False)

akroshg commented on lib/Parser/Parse.cpp:
```diff
@@ -2435,12 +2435,37 @@ bool Parser::IsImportOrExportStatementValidHere()
+template<bool buildAST> ParseNodePtr Parser::ParseImportCall()
+{
+    m_pscan->Scan();
+    ParseNodePtr specifier = ParseExpr<buildAST>(koplCma, nullptr, /* fAllowIn */FALSE, /* fAllowEllipsis */FALSE);
```
>ParseExpr [](start = 29, length = 9)

Clarification : will this function stop after first comma?

suwc replied:
Yes, it stops at ',' and doesn't accept '...'

---
In reply to: [115330698](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115330698) [](ancestors = 115330698)

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,91 @@ namespace Js
+            if (errorObject != nullptr && Js::JavascriptError::Is(errorObject))
+            {
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, errorObject, scriptContext);
+            }
```
don't we need to rethrow if above checks fail?

suwc replied:
Changed to assert + internal error.

---
In reply to: [115331615](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115331615) [](ancestors = 115331615)

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,91 @@ namespace Js
+        }
+        else
+        {
+            DWORD_PTR dwReferencingSourceContext = parentFuncBody->GetFunctionInfo()->GetFunctionProxy()->GetSourceContextInfo()->dwHostSourceContext;
```
what if the parent function is an eval? the dwHostSourceContext will be -1 in that case?

suwc replied:
Case handled in latest iteration. Added test.

---
In reply to: [115331981](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115331981) [](ancestors = 115331981)

boingoing commented on test/es6/dynamic-module-import-syntax.js:
```diff
@@ -0,0 +1,83 @@
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+// ES6 Generators Syntax tests -- verifies function* and yield syntax spec conformance
```
Nitpick, this comment doesn't look right.

suwc replied:
This file was redundant and was removed.

---
In reply to: [115339634](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115339634) [](ancestors = 115339634)

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -1863,4 +1863,16 @@ namespace Js
     }
+
+    //static
+    JavascriptPromise* JavascriptPromise::CreateEnginePromise(ScriptContext *scriptContext)
```
Cool 👍

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -220,6 +224,38 @@ namespace Js
+
+            if (this->errorObject != nullptr)
+            {
+                SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, this->GetErrorObject(), this->GetScriptContext(), this);
```
>this->GetScriptContext() [](start = 107, length = 24)

nit: just scriptContext, as we have got that in the local variable?

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -742,7 +801,15 @@ namespace Js
         if (this->errorObject != nullptr)
         {
-            JavascriptExceptionOperators::Throw(errorObject, scriptContext);
+            if (this->promise != nullptr)
```
can this codepath hit? there is an assert above which says Assert(this->errorObject == nullptr); ?

suwc replied:
Moved assert to below in the latest iteration.

---
In reply to: [115362106](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115362106) [](ancestors = 115362106)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -765,7 +832,17 @@ namespace Js
+        AUTO_NESTED_HANDLED_EXCEPTION_TYPE((ExceptionType)(ExceptionType_OutOfMemory | ExceptionType_JavascriptException));
+        ENTER_SCRIPT_IF(scriptContext, true, false, false, !scriptContext->GetThreadContext()->IsScriptActive(),
+        {
+            ret = rootFunction->CallRootFunction(outArgs, scriptContext, true);
```
If this throws exception will the promise reject properly?

suwc replied:
Added promise reject in latest iteration.

---
In reply to: [115362554](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115362554) [](ancestors = 115362554)

akroshg commented on test/es6/dynamic-module-import-syntax.js:
```diff
@@ -0,0 +1,83 @@
+        body: function () {
+                let functionBody =
+                    `
+                    assert.doesNotThrow(function () { eval("import(undefined)"); }, "undefined");
```
aren't these tests repeated from import-specifier.js file?

suwc replied:
They are not exactly the same. Import-specifier.js tests ToString operation applied on specifier. This one tests parser.

---
In reply to: [115364519](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115364519) [](ancestors = 115364519)

akroshg replied:
I don't see this file is added to the rlexe.xml, am I missing something?

---
In reply to: [115372781](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115372781) [](ancestors = 115372781,115364519)

akroshg commented on test/es6/dynamic-module-import-specifier.js:
```diff
@@ -0,0 +1,142 @@
+                            return import(${specifier});
+                        },
+                        (v)=>{
+                            assert.isTrue(false, 'Expected: promise rejected; actual: promise resolved: ' + '${message}');
```
nit: change this to assert.fail() .

suwc replied:
Done.

---
In reply to: [115364860](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115364860) [](ancestors = 115364860)

akroshg commented on test/es6/dynamic-module-functionality.js:
```diff
@@ -0,0 +1,359 @@
+    },
+];
+
+testRunner.runTests(tests, { verbose: WScript.Arguments[0] != "summary" });
```
Is there a test case which covers the nested dynamic imports ?

suwc replied:
Nest import syntax is test in dynamic-module-import-specifier.js


---
In reply to: [115367185](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115367185) [](ancestors = 115367185)

tcare commented on lib/Parser/Parse.cpp:
```diff
@@ -2435,12 +2435,37 @@ bool Parser::IsImportOrExportStatementValidHere()
 
+    RestorePoint parsedImport;
+    m_pscan->Capture(&parsedImport);
+    m_pscan->Scan();
```
Do we need to do a rewind? We are just going to scan again a few lines down anyway.

suwc replied:
This is just to make sure col# reported by error msg is not off due to the premature scan.

---
In reply to: [115371147](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115371147) [](ancestors = 115371147)

suwc commented:
>         case JsModuleHostInfo_NotifyModuleReadyCallback:

Done.

---
In reply to: [299962864](https://github.com/Microsoft/ChakraCore/pull/2913#issuecomment-299962864) [](ancestors = 299962864)

---
Refers to: lib/Jsrt/Core/JsrtCore.cpp:209 in db8d7fd. [](commit_id = db8d7fdc5b27af9f9814887a781d6e8c924e07b7, deletion_comment = False)

suwc commented:
>         if ((this->GetGrfscr() & fscrIsModuleCode) == fscrIsModuleCode || this->GetModuleID() == kmodGlobal)

Done.

---
In reply to: [299963481](https://github.com/Microsoft/ChakraCore/pull/2913#issuecomment-299963481) [](ancestors = 299963481)

---
Refers to: lib/Runtime/Base/FunctionBody.cpp:3166 in db8d7fd. [](commit_id = db8d7fdc5b27af9f9814887a781d6e8c924e07b7, deletion_comment = False)

akroshg commented on lib/Jsrt/Core/JsrtContextCore.cpp:
```diff
@@ -110,7 +129,7 @@ HRESULT ChakraCoreHostScriptContext::FetchImportedModule(Js::ModuleRecordBase* r
     {
         AUTO_NO_EXCEPTION_REGION;
-        JsErrorCode errorCode = fetchImportedModuleCallback(referencingModule, specifierVar, &dependentRecord);
+        JsErrorCode errorCode = fetch(specifierVar, &dependentRecord);//fetchImportedModuleCallback(referencingModule, specifierVar, &dependentRecord);
```
nitpick : comments can be removed

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,115 @@ namespace Js
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, errorObject, scriptContext);
+            }
+
+            Throw::InternalError();
```
what happens if 
import( { toString : function() { 
throw 1;
}});

suwc replied:
I think the reasonable thing to expect is to reject the promise with 1.

Here spec says:
IfAbruptRejectPromise(specifierString, promiseCapability).

which is questionable because specifierString will be nullptr.
I will seek clarification.


---
In reply to: [115379478](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115379478) [](ancestors = 115379478)

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,115 @@ namespace Js
+
+        // Walk the call stack if caller function is neither module code nor having host source context
+        DWORD_PTR dwReferencingSourceContext = dwReferencingSourceContext = parentFuncBody->GetFunctionInfo()->GetFunctionProxy()->GetSourceContextInfo()->dwHostSourceContext;
+        if (!parentFuncBody->IsES6ModuleCode() && dwReferencingSourceContext == Js::Constants::NoHostSourceContext)
```
if dwReferencingSourceContext == Js::Constants::NoHostSourceContext
could you not use callerUtf8SourceInfo (defined in the utf8sourceinfo)? Confirm with @agarwal.sandeep@microsoft.com that this is populated in the non-debug scenario as well.

agarwal-sandeep replied:
callerUtf8SourceInfo is populated in non-debug scenario as well. You will still need to walk the chain to get to first non evaled parent.

suwc replied:
Yes it works for a single eval case, but fails for nested eval.

---
In reply to: [115380766](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115380766) [](ancestors = 115380766)

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,115 @@ namespace Js
+                    JavascriptError::SetErrorMessageProperties(error, hr, moduleName, scriptContext);
+                    return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, error, scriptContext);
+                }
+            }
```
would be nice if this code moved to SoureTextModuleRecord. I don't think javascriptoperator.cpp needs to know any of this.

suwc replied:
Good point. Moved and renamed the function.

---
In reply to: [115381049](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115381049) [](ancestors = 115381049)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -175,6 +175,12 @@ namespace Js
+            if (this->promise != nullptr)
+            {
+                SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, this->GetErrorObject(), this->GetScriptContext(), this);
+                this->SetPromise(nullptr);
```
>SetPromise [](start = 22, length = 10)

would it be wise to reset the promise to null first and do the ResolveOrReject.. ?

suwc replied:
In fact ResolveOrRejectDynamicImportPromise() takes care of this already. Removing this line and in similar cases

---
In reply to: [115382160](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115382160) [](ancestors = 115382160)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -175,6 +175,12 @@ namespace Js
             OUTPUT_TRACE_DEBUGONLY(Js::ModulePhase, _u("\t>NotifyParentAsNeeded\n"));
+            if (this->promise != nullptr)
+            {
+                SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, this->GetErrorObject(), this->GetScriptContext(), this);
```
>this->GetScriptContext() [](start = 107, length = 24)

nitpick : use the local scriptContext (also you can use the errorObject as well)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -220,6 +224,38 @@ namespace Js
+
+            if (this->errorObject != nullptr)
+            {
+                SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, this->GetErrorObject(), scriptContext, this);
```
>scriptContext [](start = 107, length = 13)

nitpick : could you not use the this->scriptContext? I am not sure why we need to have local variable for a scriptcontext.

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -765,7 +834,41 @@ namespace Js
+            AssertMsg(errorObject != nullptr && Js::JavascriptError::Is(errorObject), "ModuleEvaluation: Unhandled exception thrown by calling root function");
+            if (errorObject != nullptr && Js::JavascriptError::Is(errorObject))
+            {
+                ResolveOrRejectDynamicImportPromise(false, errorObject, scriptContext);
```
setPromise to null?

suwc replied:
Adding 'this' as last param and SetPromise(nullptr) will be taken care of.

---
In reply to: [115383242](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115383242) [](ancestors = 115383242)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -765,7 +834,41 @@ namespace Js
+            Js::JavascriptError * error = scriptContext->GetLibrary()->CreateError();
+            JavascriptError::SetErrorMessageProperties(error, E_OUTOFMEMORY, this->GetSpecifierSz(), scriptContext);
+            ResolveOrRejectDynamicImportPromise(false, error, scriptContext);
+            return scriptContext->GetLibrary()->GetUndefined();
```
we are returning undefined in the OOM case? why not re-thorw?

suwc replied:
Only in the dynamic import case, we reject the promise with error object and return undefined.
In the static case, we should throw the error object. Making the change.

---
In reply to: [115383826](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115383826) [](ancestors = 115383826)

suwc commented:
@dotnet-bot 
run Windows x64_release

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,104 @@ namespace Js
+                    if (walker.GetCaller(&caller) && caller != nullptr && caller->IsScriptFunction())
+                    {
+                        parentFuncBody = caller->GetFunctionBody();
+                        dwReferencingSourceContext = dwReferencingSourceContext = parentFuncBody->GetFunctionInfo()->GetFunctionProxy()->GetSourceContextInfo()->dwHostSourceContext;
```
nit: function body has function GetHostSourceContext(), which will give you the dwHostSOurceContext.

suwc commented:
@dotnet-bot 
test Windows x64_release

akroshg commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -9639,6 +9639,104 @@ namespace Js
+
+                do
+                {
+                    if (walker.GetCaller(&caller) && caller != nullptr && caller->IsScriptFunction())
```
interesting to see what happens when the import is called in the setTimeout. Will you add a test case just to see that we don't get the fatal error.

akroshg replied:
I couldn't find a scenario which involves the nested imports also dynamic->static->dynamic. sort of?

---
In reply to: [115561112](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115561112) [](ancestors = 115561112)

suwc replied:
I am still working on these test cases. They will come in later.

---
In reply to: [115577016](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r115577016) [](ancestors = 115577016,115561112)

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -765,7 +844,52 @@ namespace Js
+                }
+            }
+
+            throw(err);
```
throw outside of catch handler. (look at other example in the code base, where they use different function for re-throw).

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -765,7 +844,52 @@ namespace Js
+            if (this->promise != nullptr)
+            {
+                Var errorObject = err.GetAndClear()->GetThrownObject(scriptContext);
+                AssertMsg(errorObject != nullptr, "ModuleEvaluation: null error object thrown from root function");
```
>AssertMsg [](start = 16, length = 9)

nit : convert to AssertOrFailFast so that you don't have to have if/else blocks

akroshg commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -739,19 +813,24 @@ namespace Js
         }
+
+        Assert(this->errorObject == nullptr);
         Assert(!WasEvaluated());
```
nit: SetWasEvaluated has this assert covered.

akroshg commented:
Is import() configurable through config flags or it is enabled by default on?

suwc commented:
We are enabling it by default.

---
In reply to: [300271647](https://github.com/Microsoft/ChakraCore/pull/2913#issuecomment-300271647) [](ancestors = 300271647)

suwc commented:
Thanks for the code review.
